middle_end/flambda/types/env/aliases.ml

```diff
@@ -19,163 +19,288 @@ module Aliases_of_canonical_element : sig
-        t1.aliases t2.aliases
+    let earliest_normal =
+      Earliest_alias.union t1.earliest_normal t2.earliest_normal
     in
-    let t =
-      { aliases;
-        all = Simple.Set.inter t1.all t2.all;
-      }
+    let earliest_ge_phantom =
+      Earliest_alias.union t1.earliest_ge_phantom t2.earliest_ge_phantom
     in
-    invariant t;
-    t
-
-  let rename rename_simple { aliases; all } =
-    let aliases =
-      Name_mode.Map.map (fun elts -> Simple.Set.map rename_simple elts)
-        aliases
+    let earliest_ge_in_types =
+      Earliest_alias.union t1.earliest_ge_in_types t2.earliest_ge_in_types
     in
+    { all; earliest; earliest_normal; earliest_ge_phantom;
+      earliest_ge_in_types; }
+
+  let disjoint t1 t2 =
+    Simple.Set.intersection_is_empty t1.all t2.all
+
+  let update_all_earliest
+        { all; earliest; earliest_normal;
+          earliest_ge_phantom; earliest_ge_in_types; } ~f =
+    let earliest = f earliest in
+    let earliest_normal = f earliest_normal in
+    let earliest_ge_phantom = f earliest_ge_phantom in
+    let earliest_ge_in_types = f earliest_ge_in_types in
+    { all; earliest; earliest_normal; earliest_ge_phantom;
+      earliest_ge_in_types; }
+
+  let rename rename_simple
+        ({ all; earliest = _; earliest_normal = _;
+           earliest_ge_phantom = _; earliest_ge_in_types = _; } as t) =
     let all = Simple.Set.map rename_simple all in
-    { aliases; all }
-
-  let merge t1 t2 =
-    let aliases =
-      Name_mode.Map.union (fun _mode set1 set2 ->
-          Some (Simple.Set.union set1 set2))
-        t1.aliases
-        t2.aliases
-    in
-    let all = Simple.Set.union t1.all t2.all in
-    { aliases; all; }
-
-  let move_variables_to_mode_in_types { aliases; all; } =
-    let (no_vars_aliases, all_variables) =
-      Name_mode.Map.fold (fun mode aliases (no_vars_aliases, all_variables) ->
-          let (vars, non_vars) = Simple.Set.partition Simple.is_var aliases in
-          let no_vars_aliases =
-            if Simple.Set.is_empty non_vars then no_vars_aliases
-            else Name_mode.Map.add mode non_vars no_vars_aliases
-          in
-          no_vars_aliases, Simple.Set.union vars all_variables)
-        aliases
-        (Name_mode.Map.empty, Simple.Set.empty)
-    in
-    let aliases =
-      if Name_mode.Map.mem Name_mode.in_types no_vars_aliases
-      then Misc.fatal_errorf "move_variables_to_mode_in_types: \
-             The following non-vars have mode In_types:@ %a"
-             Simple.Set.print
-             (Name_mode.Map.find Name_mode.in_types no_vars_aliases)
-      else
-        if Simple.Set.is_empty all_variables then no_vars_aliases
-        else Name_mode.Map.add Name_mode.in_types all_variables no_vars_aliases
+    let t =
+      update_all_earliest t ~f:(Earliest_alias.map_name ~f:rename_simple)
     in
-    { aliases; all; }
+    { t with all; }
+
+  let move_variables_to_mode_in_types t =
+    update_all_earliest t ~f:(fun earliest ->
```

Comment:
This looks wrong. I would expect `earliest` and `earliest_ge_in_types` to be left unchanged by this function.

Reply:
(Although #393 gets rid of this function completely, so maybe it isn't that important anyway)

Reply:
!! Yes, of course.